### ISSUE TITLE
Workspace计算溢出

### ISSUE THREAD

**Issue by WuTYSFG** [WuTYSFG/triton-ascend#8]
### Describe the bug

uint64_t totalWorkSpaceSize = {workspace_size} * blockNum4Workspace;
{workspace_size} 是 Python 端 int(metadata.workspace_size) 注入的 C++ int 字面量 ， blockNum4Workspace 是 uint32_t 。
C++ 隐式转换规则里，int 遇上 uint32_t 时两边都被提升为 uint32_t ， 乘法在 uint32 下完成 ，结果先按 2^32 取模， 然后再拓宽到 uint64_t赋值
当workspace_size * blockNum4Workspace > 2^32 ，结果就会被截断，再赋给 uint64 已经是错误值
65536=2^16，uint32_t 的最大值是 2^32 − 1，65536*65536=2^32高位发生截断导致计算结果等于0

### Environment details

NPU

**Comment by WuTYSFG** [comment]
Workspace 是 Ascend 后端模拟 CUDA cooperative launch 的机制 ：CUDA 上同一个 kernel 的所有 CTA 可以直接访问一个跨 CTA 共享的临时 buffer（cooperative groups + 显式 scratch）；Ascend 没有等价特性，Triton-Ascend 因此为每个 kernel 预留一块 GM区域 作为 scratch space，由 launcher 在每次启动时统一申请。

**Comment by WuTYSFG** [comment]
### 设计动机
CUDA 等价能力有三种，Ascend 都没有：

| CUDA 提供 Ascend | 替代方案 |
| --- | --- |
| cooperative_launch 跨 CTA 同步 | lock_num （lock数组） |
|  __cluster_shared_memory 跨 CTA 共享 | buffer workspace （per-block GM scratch） |
| cudaMallocAsync 流上临时分配 | workspace（kernel启动时一次性预分配）|

### Workspace 的具体用途
- cross-CTA 同步的状态存放 ： al.sync_block_set/wait 之间的 fence 状态、event counter
- 超出 UB 容量的中间 buffer ：单个 AI Core 的 UB 只有 ~256 KB，超出的 tiling/multi-buffer staging 会 spill 到 workspace
- debug/profiling 数据 ：device print、profiler 回调的临时缓冲
- SDPA / attention 类大算子 的 softmax 统计（ m_i / l_i ）、reduction 中间结果

### workspace_size 是怎么算出来的
用户写 Triton 代码时不能直接指定 workspace_size ——它由 MLIR lowering pass 自动推断：
```
@triton.jit kernel
 ↓ triton-compile → TTIR
 ↓ triton-ascend pipeline ↓ hivm-cube / hivm-vector 方言
 ↓ BiShengIR HIVM pass: insert-infer-workspace-size-func
 ↓ scan all memref.alloc ops, compute max cross-block footprint
 ↓ metadata.workspace_size = inferred_bytes
```

### 触发 workspace_size > 0 的 Triton 写法
不是所有 Triton kernel 都会触发。要触发就必须让 MLIR pass 看到"必须跨 block 可见的 memref.alloc"。常见路径：
| Triton 写法 | 触发原因 |
| --- | --- |
| al.sync_block_set/wait cube↔vector | cross-CTA 同步状态 |
| 大 matmul（block ≥ 128×128） | multi-buffer 流水线 staging |
| tl.softmax + 大 dim | reduction 中间 buffer跨 block |
| SDPA 风格 attention | QK^T + softmax 输出 |

**Comment by WuTYSFG** [comment]
```python
#!/usr/bin/env python3
# Copyright (c) Huawei Technologies Co., Ltd. 2026. All rights reserved.
"""End-to-end test: compile a real Triton kernel that triggers
``metadata.workspace_size > 0`` via the cross-CTA sync path, then run
it on the (mocked) Ascend launcher to confirm workspace allocation is
requested.

Run with::

    pytest test_workspace_real_kernel.py -v
"""
import importlib.util
import os
from pathlib import Path
from types import SimpleNamespace
from unittest.mock import MagicMock, patch

import pytest

import triton
import triton.language as tl
import triton.language.extra.cann.extension as al

# ---------------------------------------------------------------------------
# Boot driver
# ---------------------------------------------------------------------------

_THIS_DIR = Path(__file__).resolve().parent
_DRIVER_PATH = _THIS_DIR.parents[1] / "backend" / "driver.py"


def _load_driver():
    spec = importlib.util.spec_from_file_location(
        "ascend_driver_under_test", _DRIVER_PATH)
    mod = importlib.util.module_from_spec(spec)
    assert spec.loader is not None
    spec.loader.exec_module(mod)
    return mod


driver = _load_driver()


# ---------------------------------------------------------------------------
# Kernel: a real, runnable Triton kernel that uses cross-CTA sync
# ---------------------------------------------------------------------------


@triton.jit
def kernel_real_workspace_demo(x_ptr, y_ptr, out_ptr, n, BLOCK: tl.constexpr):
    """Cube computes a partial sum, signals Vector, Vector finalises.

    The ``al.sync_block_set`` / ``al.sync_block_wait`` pair forces the
    BiShengIR lowering to allocate cross-block sync state, which the
    ``insert-infer-workspace-size-func`` pass counts into
    ``metadata.workspace_size``.

    This kernel is functionally correct on a real device — it adds a
    scalar (computed on Cube) to every element of ``y_ptr`` and writes
    to ``out_ptr``. The demo value is irrelevant; what matters is that
    compilation produces a non-zero ``workspace_size``.
    """
    pid = tl.program_id(0)
    i = pid * BLOCK + tl.arange(0, BLOCK)

    # ---- Cube stage: load x, reduce, signal vector -----------------
    with al.scope("cube"):
        x = tl.load(x_ptr + i, mask=i < n)
        scalar = tl.sum(x, axis=0)
        al.sync_block_set("cube", "vector", event_id=0)

    # ---- Vector stage: wait, add scalar to y, store ---------------
    with al.scope("vector"):
        al.sync_block_wait("cube", "vector", event_id=0)
        y = tl.load(y_ptr + i, mask=i < n)
        out = y + scalar
        tl.store(out_ptr + i, out, mask=i < n)


@triton.jit
def kernel_no_workspace_baseline(x_ptr, y_ptr, out_ptr, n, BLOCK: tl.constexpr):
    """Pure elementwise kernel — no cross-CTA sync. Should compile
    with ``metadata.workspace_size == 0`` for contrast.
    """
    i = tl.program_id(0) * BLOCK + tl.arange(0, BLOCK)
    x = tl.load(x_ptr + i, mask=i < n)
    y = tl.load(y_ptr + i, mask=i < n)
    tl.store(out_ptr + i, x + y, mask=i < n)


# ---------------------------------------------------------------------------
# Real Triton compile + metadata inspection
# ---------------------------------------------------------------------------


def _compile_real_kernel(kernel, workspace_size_expected=None):
    """Run the real ``triton.compile`` flow against the kernel and
    return the ``compiled_kernel.metadata`` object.

    Uses the same API the user-facing compiler exposes, so any
    ``metadata.workspace_size`` value observed here is exactly what the
    driver sees at launch time.
    """
    import torch

    # Ascend target — use the same canonical string Triton's docs use
    target = triton.runtime.driver.active.get_current_target()

    x = torch.empty(64, dtype=torch.float32)
    compiled = triton.compile(
        kernel,
        signature={
            "x_ptr": "*fp32",
            "y_ptr": "*fp32",
            "out_ptr": "*fp32",
            "n": "i32",
        },
        constants={"BLOCK": 16},
        target=target,
    )
    return compiled


def test_baseline_kernel_has_zero_workspace():
    """Sanity: a kernel that does not use cross-CTA sync compiles with
    ``metadata.workspace_size == 0``.
    """
    compiled = _compile_real_kernel(kernel_no_workspace_baseline)
    md = compiled.metadata
    ws = getattr(md, "workspace_size", 0) or 0
    assert ws == 0, (
        f"baseline elementwise kernel should have workspace_size=0, "
        f"got {ws}")


def test_sync_kernel_triggers_workspace_metadata():
    """Main check: a kernel with ``al.sync_block_set/wait`` triggers
    the BiShengIR pass to set a non-zero ``metadata.workspace_size``.
    """
    try:
        compiled = _compile_real_kernel(kernel_real_workspace_demo)
    except Exception as ex:
        pytest.skip(
            "Could not compile real kernel on this host "
            f"(no Ascend target available?): {ex}")

    md = compiled.metadata
    ws = getattr(md, "workspace_size", None)
    if ws is None:
        pytest.skip("Compiled kernel metadata lacks workspace_size field")

    assert ws > 0, (
        f"kernel_real_workspace_demo should infer workspace_size > 0, "
        f"got {ws}. Check that the MLIR pass "
        "insert-infer-workspace-size-func is being run.")


def test_driver_launcher_allocates_workspace_for_sync_kernel():
    """Combined check: compile the sync kernel, then drive the
    ``NPULauncher`` with mocked device layer to confirm the launcher
    actually calls ``allocate_memory`` with the inferred size.
    """
    try:
        compiled = _compile_real_kernel(kernel_real_workspace_demo)
    except Exception as ex:
        pytest.skip(f"Could not compile real kernel: {ex}")

    md = compiled.metadata
    ws = getattr(md, "workspace_size", 0) or 0
    if ws == 0:
        pytest.skip("compiled metadata workspace_size is 0")

    # Patch every side-effectful piece so the test runs on any host
    allocator_mock = MagicMock()
    allocate_mock = MagicMock(return_value=object())
    allocator_mock.get.return_value = allocate_mock
    launch_mock = MagicMock(return_value=0)

    fake_module = SimpleNamespace(launch=launch_mock)
    fake_spec = SimpleNamespace(loader=SimpleNamespace(
        exec_module=lambda m: None))

    monkey = pytest.MonkeyPatch()
    try:
        monkey.setenv("TRITON_COMPILE_ONLY", "true")  # skip real launch
        monkey.setattr(driver._allocation, "_allocator", allocator_mock)
        with patch.object(driver.NPULauncher,
                          "_make_launcher_stub_path",
                          return_value="/tmp/fake_launcher.so"), \
             patch("importlib.util.spec_from_file_location",
                    return_value=fake_spec), \
             patch("importlib.util.module_from_spec",
                    return_value=fake_module):
            # Build launcher with the *real* compiled metadata
            launcher = driver.NPULauncher(compiled.src, md)
            launcher(8, 1, 1, 0, None, {"hash": "h"},
                     None, None, None)
    finally:
        monkey.undo()

    # The C launcher invokes allocate_memory at the workspace path.
    # We can verify by re-rendering the source for the same metadata
    # and grepping for the workspace block.
    src = driver.make_launcher(
        compiled.src.constants if hasattr(compiled.src, "constants") else {},
        compiled.src.signature,
        md,
    )
    assert "totalWorkSpaceSize" in src, (
        "real sync kernel metadata should produce workspace block in launcher")
    assert f"= (uint64_t){ws} * blockNum4Workspace" in src or \
        f"= {ws} * blockNum4Workspace" in src


# ---------------------------------------------------------------------------
# Manual demo
# ---------------------------------------------------------------------------


if __name__ == "__main__":
    print("=" * 60)
    print("kernel_real_workspace_demo - compiled metadata")
    print("=" * 60)
    try:
        compiled = _compile_real_kernel(kernel_real_workspace_demo)
        md = compiled.metadata
        ws = getattr(md, "workspace_size", None)
        print(f"metadata.workspace_size = {ws}")
        print(f"lock_num = {getattr(md, 'lock_num', None)}")
    except Exception as ex:
        print(f"compile failed: {ex}")

    print()
    print("=" * 60)
    print("kernel_no_workspace_baseline - compiled metadata")
    print("=" * 60)
    try:
        compiled = _compile_real_kernel(kernel_no_workspace_baseline)
        md = compiled.metadata
        ws = getattr(md, "workspace_size", None)
        print(f"metadata.workspace_size = {ws}")
    except Exception as ex:
        print(f"compile failed: {ex}")
```

**Comment by WuTYSFG** [comment]
```python
@triton.jit
def matmul_mul_kernel(
    A_ptr, B_ptr, C_ptr, Out_ptr,
    M, N, K,
    stride_am, stride_ak,
    stride_bk, stride_bn,
    stride_cm, stride_cn,
    stride_om, stride_on,
    BLOCK_M: tl.constexpr,
    BLOCK_N: tl.constexpr,
    BLOCK_K: tl.constexpr,
):
    pid_m = tl.program_id(0)
    pid_n = tl.program_id(1)

    rm = pid_m * BLOCK_M + tl.arange(0, BLOCK_M)
    rn = pid_n * BLOCK_N + tl.arange(0, BLOCK_N)
    rk = tl.arange(0, BLOCK_K)

    m_mask = rm[:, None] < M
    n_mask = rn[None, :] < N
    full_mask = m_mask & n_mask

    A = tl.load(
        A_ptr + rm[:, None] * stride_am + rk[None, :] * stride_ak,
        mask=m_mask, other=0.0)

    B = tl.load(
        B_ptr + rk[:, None] * stride_bk + rn[None, :] * stride_bn,
        mask=n_mask, other=0.0)

    C = tl.load(
        C_ptr + rm[:, None] * stride_cm + rn[None, :] * stride_cn,
        mask=full_mask, other=0.0)

    AB = tl.dot(A, B)
    Out = AB * C
    tl.store(
        Out_ptr + rm[:, None] * stride_om + rn[None, :] * stride_on,
        Out, mask=full_mask)


def _make_inputs(M: int, N: int, K: int, dtype=torch.float32, device="cpu"):
    """Build a packed set of A, B, C, Out on the given device."""
    torch.manual_seed(0)
    A = torch.randn(M, K, dtype=dtype, device=device)
    B = torch.randn(K, N, dtype=dtype, device=device)
    C = torch.randn(M, N, dtype=dtype, device=device)
    Out = torch.empty(M, N, dtype=dtype, device=device)
    return A, B, C, Out


def _ref(A, B, C):
    return torch.matmul(A, B) * C


def _launch_kernel(A, B, C, Out, *, BLOCK_M=32, BLOCK_N=32, BLOCK_K=32):
    M, K = A.shape
    _, N = B.shape
    grid = (triton.cdiv(M, BLOCK_M), triton.cdiv(N, BLOCK_N))
    matmul_mul_kernel[grid](
        A, B, C, Out,
        M, N, K,
        A.stride(0), A.stride(1),
        B.stride(0), B.stride(1),
        C.stride(0), C.stride(1),
        Out.stride(0), Out.stride(1),
        BLOCK_M=BLOCK_M, BLOCK_N=BLOCK_N, BLOCK_K=BLOCK_K,
    )
```

**Comment by WuTYSFG** [comment]
fix pr: https://github.com/triton-lang/triton-ascend/pull/1714
